File: day20/task1.zig
const std = @import("std");
const uint = u64;

fn housePresents(house: uint) uint {
  var total = @as(uint, 0);
  var i = @as(uint, 1);
  while (i * i <= house) {
    if (house % i == 0) {
      total += i;
      if (i * i != house) {
        total += house / i;
      }
    }
    i += 1;
  }
  return total;
}

pub fn main() !void {
  const limit = 3310000;
  var house = @as(uint, 1);
  while (housePresents(house) < limit) {
    house += 1;
  }
  std.debug.print("Answer: {}\n", .{house});
}

File: day20/task1.R
library(tidyverse)

house_presents <- function(house) {
  seq_len(house) %>%
    (\(elves) elves[house %% elves == 0] * 10) %>%
    sum
}

find_house <- function(limit) {
  house <- 1
  while (TRUE) {
    presents <- house_presents(house)
    if (presents >= limit) {
      return(house)
    }
    paste0("House #", house, ": ", presents) |> print()
    house <- house + 1
  }
}

limit <- 33100000
find_house(limit)
